---
output:
    github_document:
        html_preview: false
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  eval = FALSE
)
```

# CAMDAC

Copy-number Aware Methylation Deconvolution Analysis of Cancer (CAMDAC) is an R library for deconvolving bulk tumor DNA methylation (bisulfite) sequencing data ([Larose Cadieux et al., 2022, bioXriv](https://www.biorxiv.org/content/10.1101/2020.11.03.366252v2)).

<!-- badges: start -->
<!-- badges: end -->

## Documentation

Visit [https://vanloo-lab.github.io/CAMDAC/](https://vanloo-lab.github.io/CAMDAC/).

## Quickstart

CAMDAC can be installed from an R console:

``` r
remotes::install_github("VanLoo-lab/CAMDAC")
```

Download reference datasets required to run CAMDAC for RRBS and/or WGBS analysis [from the Zenodo repository: (10565423)](https://zenodo.org/records/10565423). An R helper function is provided for convenience:

```{r}
CAMDAC::download_pipeline_files(bsseq = "rrbs", directory = "./refs")
CAMDAC::download_pipeline_files(bsseq = "wgbs", directory = "./refs")
```

Run the tumor-normal deconvolution pipeline with test data:

> [!NOTE]  
> We provide downsampled BAM files for testing the pipeline. For representative results, please use your own BAM files.

```{r example}
library(CAMDAC)

tumor_bam <- system.file("testdata", "tumor.bam", package = "CAMDAC")
normal_bam <- system.file("testdata", "normal.bam", package = "CAMDAC")

# Select samples for basic tumor-normal analysis
tumor <- CamSample(id = "T", sex = "XY", bam = tumor_bam)
normal <- CamSample(id = "N", sex = "XY", bam = normal_bam)

# Configure pipeline
config <- CamConfig(
  outdir = "./results", bsseq = "rrbs", lib = "pe",
  build = "hg38", refs = "./refs", n_cores = 1, cna_caller='ascat'
)

# Run CAMDAC
CAMDAC::pipeline(
  tumor, germline = normal, infiltrates = normal, origin = normal, config
)
```

For a more detailed walkthrough with test data, see `vignette("pipeline")`.

## Contributing

To contribute to CAMDAC, simply fork the repository and submit a merge request.

We use the following commands to check the package before commits:

```{r}
library(devtools)

# Install dev dependencies
devtools::install_dev_deps("VanLoo-lab/CAMDAC")

# Update docs
devtools::document()

# Run tests
devtools::test()

# Build readme
rmarkdown::render('README.Rmd', output_format='github_document', output_file='README.md')

# Check package builds
devtools::check()

# Build documentation
pkgdown::build_articles()
pkgdown::build_site(examples=FALSE, devel=TRUE, lazy=TRUE, preview=FALSE)
pkgdown::preview_site() # To view. Or: python3 -m http.server --directory docs 8000
```
