<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>DNA methylation panels â€¢ CAMDAC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="DNA methylation panels">
<meta property="og:description" content="CAMDAC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CAMDAC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/setup.html">Installation</a>
    </li>
    <li>
      <a href="../articles/pipeline.html">CAMDAC pipeline</a>
    </li>
    <li>
      <a href="../articles/output.html">Results</a>
    </li>
    <li>
      <a href="../articles/visualisation.html">Visualisation</a>
    </li>
    <li>
      <a href="../articles/questions.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/panels.html">DNA methylation panels</a>
    </li>
    <li>
      <a href="../articles/copy_number_alt.html">Alternative CNA calling</a>
    </li>
    <li>
      <a href="../articles/allele-specific.html">Allele-specific Methylation</a>
    </li>
    <li>
      <a href="../articles/deconvolution.html">CAMDAC Deconvolution Only</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>DNA methylation panels</h1>
            
      
      
      <div class="hidden name"><code>panels.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="panel-deconvolution">Panel deconvolution<a class="anchor" aria-label="anchor" href="#panel-deconvolution"></a>
</h2>
<div class="section level3">
<h3 id="create-a-methylation-panel-from-multiple-normal-bam-files">Create a methylation panel from multiple normal BAM files<a class="anchor" aria-label="anchor" href="#create-a-methylation-panel-from-multiple-normal-bam-files"></a>
</h3>
<p>CAMDAC supports the use of multiple DNA methylation BAM files as a source of the normal infiltrates or normal cell of origin.</p>
<p>To create a panel, process your BAM files with the CAMDAC allele counter:</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CAMDAC</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get BAM files</span></span>
<span><span class="va">b_normal1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"inst/testdata/normal.bam"</span><span class="op">)</span></span>
<span><span class="va">b_normal2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"inst/testdata/normal.bam"</span><span class="op">)</span></span>
<span><span class="va">b_normal3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"inst/testdata/normal.bam"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run allele counter</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">file</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">b_normal1</span>, <span class="va">b_normal2</span>, <span class="va">b_normal3</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">prefix</span> <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_file.html" class="external-link">path_ext_remove</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span></span>
<span>    <span class="va">outfile</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">prefix</span>, <span class="st">".all.SNPs.CG.csv.gz"</span><span class="op">)</span></span>
<span>    <span class="va">data</span> <span class="op">=</span> <span class="fu"><a href="../reference/cmain_count_alleles.html">cmain_count_alleles</a></span><span class="op">(</span><span class="va">bam_file</span><span class="op">)</span></span>
<span>    <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fwrite.html" class="external-link">fwrite</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">outfile</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
<p>The allele counts files can then be merged into a single file for the panel containing methylation data for deconvolution:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">panel_counts</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="st">"."</span>, glob<span class="op">=</span><span class="st">"*.SNPs.CG.csv.gz"</span><span class="op">)</span></span>
<span><span class="va">panel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/panel_meth_from_counts.html">panel_meth_from_counts</a></span><span class="op">(</span><span class="va">panel_counts</span><span class="op">)</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fwrite.html" class="external-link">fwrite</a></span><span class="op">(</span><span class="va">panel</span>, <span class="st">"panel.m.csv.gz"</span><span class="op">)</span></span></code></pre></div>
<p>By default, panel counts are merged by summing the methylation read counts for each CpG site. You can customise the proportion of each sample that is used in the panel by specifying the <code>ac_props</code> argument in panel_meth_from_counts. To get the mean across each CpG site, simply pass equal proportions for each sample.</p>
<p>To run CAMDAC with your newly created panel, attach your panel to a CamSample object using the <code>meth</code> argument.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load test data</span></span>
<span><span class="va">b_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"tumor.bam"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="va">b_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"normal.bam"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Setup CAMDAC samples</span></span>
<span><span class="va">tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"tumor"</span>, sex<span class="op">=</span><span class="st">"XY"</span>, bam<span class="op">=</span><span class="va">b_tumor</span><span class="op">)</span></span>
<span><span class="va">normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"normal"</span>, sex<span class="op">=</span><span class="st">"XY"</span>, bam<span class="op">=</span><span class="va">b_normal</span><span class="op">)</span></span>
<span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamConfig.html">CamConfig</a></span><span class="op">(</span>outdir<span class="op">=</span><span class="st">"./results"</span>, ref<span class="op">=</span><span class="st">"./pipeline_files"</span>, bsseq<span class="op">=</span><span class="st">"wgbs"</span>, lib<span class="op">=</span><span class="st">"pe"</span>, cores<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Setup panel sample</span></span>
<span><span class="va">panel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"panel"</span>, sex<span class="op">=</span><span class="st">"XY"</span><span class="op">)</span></span>
<span><span class="va">panel_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"test_panel.m.csv.gz"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/attach_output.html">attach_output</a></span><span class="op">(</span><span class="va">panel</span>, <span class="va">config</span>, <span class="st">"meth"</span>, <span class="va">panel_file</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run CAMDAC with panel</span></span>
<span><span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span></span>
<span>    tumor<span class="op">=</span><span class="va">tumor</span>,</span>
<span>    germline<span class="op">=</span><span class="va">normal</span>,</span>
<span>    infiltrates<span class="op">=</span><span class="va">panel</span>,</span>
<span>    origin<span class="op">=</span><span class="va">panel</span>,</span>
<span>    config<span class="op">=</span><span class="va">config</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-a-methylation-panel-from-a-matrix-of-beta-values">Create a methylation panel from a matrix of beta values<a class="anchor" aria-label="anchor" href="#create-a-methylation-panel-from-a-matrix-of-beta-values"></a>
</h3>
<p>If you have not started from BAM files, you can create a panel using a matrix of beta values:</p>
<table class="table">
<thead><tr class="header">
<th>sample1</th>
<th>sample2</th>
<th>sample3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>0.5</td>
<td>0.6</td>
<td>0.7</td>
</tr>
<tr class="even">
<td>0.4</td>
<td>0.5</td>
<td>0.6</td>
</tr>
</tbody>
</table>
<p>Additionally, a data frame specifying the positions of each CpG site in the beta value matrix is required. Here, start and end refer to the C and G of the CpG site respectively:</p>
<table class="table">
<thead><tr class="header">
<th>chrom</th>
<th>start</th>
<th>end</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>chr1</td>
<td>100</td>
<td>101</td>
</tr>
<tr class="even">
<td>chr1</td>
<td>200</td>
<td>201</td>
</tr>
</tbody>
</table>
<p>The matrix and CpG locations can be passed directly to the <code><a href="../reference/panel_meth_from_beta.html">panel_meth_from_beta()</a></code> function, along with settings.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load beta values and chromosome positions</span></span>
<span><span class="va">ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"test_panel_from_beta.csv"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">ex</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">=</span> <span class="va">data</span><span class="op">[</span>, <span class="fl">4</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">]</span> <span class="co"># Beta value matrix with 3 samples</span></span>
<span></span>
<span><span class="co"># Create panel from beta values</span></span>
<span><span class="va">panel_beta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/panel_meth_from_beta.html">panel_meth_from_beta</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">mat</span>,</span>
<span>  chrom <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">chrom</span>,</span>
<span>  start <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">start</span>,</span>
<span>  end <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">end</span>, </span>
<span>  cov <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  props <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.8</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="co"># Proportions of each sample in panel</span></span>
<span>  min_samples <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  max_sd <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As CAMDAC requires coverage at each CpG site to estimate uncertainty, the <code>cov</code> value is given to all CpG sites when building a panel from beta values. Additionally, if any beta values are missing from a sample, proportions are recalculated among the remaining samples as this is the only information available to build the panel for that site.</p>
<p>There are two experimental arguments that can be set to filter CpG sites from the panel:</p>
<ul>
<li><p>min_samples: The minimum number of samples that have to have a beta value for a CpG to be included in the panel. The idea here is if you have sparse data, you can skip sites where you arenâ€™t confident in the panel. Set this to 1 to use any sample.</p></li>
<li><p>max_sd: Maximum standard deviation of beta values across samples a CpG must have to be included in the panel. The idea here is that when combining many bulk methylomes from the same tissue, sites with high variability reflect sample-specific differences and their averages are less reliable for use in a methylation panel.</p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Elizabeth Larose Cadieux, Nana Mensah, Carla Castignani, Jonas Demeulemeester, Peter Van Loo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
