<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>FAQs • CAMDAC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="FAQs">
<meta property="og:description" content="CAMDAC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CAMDAC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/setup.html">Installation</a>
    </li>
    <li>
      <a href="../articles/pipeline.html">CAMDAC pipeline</a>
    </li>
    <li>
      <a href="../articles/output.html">Results</a>
    </li>
    <li>
      <a href="../articles/visualisation.html">Visualisation</a>
    </li>
    <li>
      <a href="../articles/questions.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/panels.html">DNA methylation panels</a>
    </li>
    <li>
      <a href="../articles/copy_number_alt.html">Alternative CNA calling</a>
    </li>
    <li>
      <a href="../articles/allele-specific.html">Allele-specific Methylation</a>
    </li>
    <li>
      <a href="../articles/deconvolution.html">CAMDAC Deconvolution Only</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>FAQs</h1>
            
      
      
      <div class="hidden name"><code>questions.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="general-faq">General FAQ<a class="anchor" aria-label="anchor" href="#general-faq"></a>
</h2>
<div class="section level3">
<h3 id="what-if-i-dont-have-a-cna-or-matched-germline-sample">What if I don’t have a CNA or matched germline sample?<a class="anchor" aria-label="anchor" href="#what-if-i-dont-have-a-cna-or-matched-germline-sample"></a>
</h3>
<p>Unfortunately, CAMDAC cannot be run without a CNA solution or a normal sample from which to derive one.</p>
</div>
<div class="section level3">
<h3 id="i-have-unmatched-normal-samples-from-multiple-patients--can-i-use-camdac-to-merge-them">I have unmatched normal samples from multiple patients. Can I use CAMDAC to merge them?<a class="anchor" aria-label="anchor" href="#i-have-unmatched-normal-samples-from-multiple-patients--can-i-use-camdac-to-merge-them"></a>
</h3>
<p>Yes, please follow see the vignette <code>vignette("panel")</code>.</p>
</div>
<div class="section level3">
<h3 id="i-want-to-run-camdac-on-something-other-than-hg19-or-hg38">I want to run CAMDAC on something other than hg19 or hg38<a class="anchor" aria-label="anchor" href="#i-want-to-run-camdac-on-something-other-than-hg19-or-hg38"></a>
</h3>
<p>Please raise an issue on GitHub to request files for a new reference genome.</p>
</div>
</div>
<div class="section level2">
<h2 id="features">Features<a class="anchor" aria-label="anchor" href="#features"></a>
</h2>
<div class="section level3">
<h3 id="can-i-skip-steps-of-the-analysis">Can I skip steps of the analysis?<a class="anchor" aria-label="anchor" href="#can-i-skip-steps-of-the-analysis"></a>
</h3>
<p>When calling <code>pipeline</code> if you do not give a normal infiltrate or cell of origin, the pipeline skips deconvolution and differential methylation respectively. This may be useful to run a quick first-pass to find and refit copy number solutions. When CAMDAC has found a solution and is rerun with the same tumor, config, and normal, the <code>infiltrates</code> and <code>cell_of_origin</code> arguments will continue the pipeline where it left off. The entire pipeline can be re-run be deleting the output directory or setting <code>overwrite=TRUE</code> in the <code>CamConfig</code>.</p>
</div>
<div class="section level3">
<h3 id="how-do-i-run-individual-steps-of-the-camdac-pipeline">How do I run individual steps of the CAMDAC pipeline?<a class="anchor" aria-label="anchor" href="#how-do-i-run-individual-steps-of-the-camdac-pipeline"></a>
</h3>
<p>The simplest way is to call <code>pipeline</code> with <code>overwrite=FALSE</code> in your config, giving the right normal sample for your step. Additionally, you <code>CamConfig</code> must run with the same output directory.</p>
<p>If for any reason, you have changed the output directory structure from previous run, you can initiate CAMDAC by manually passing outputs to <code>CamSample</code> objects. See the vignette <code><a href="../articles/output.html">vignette("output")</a></code> for more information.</p>
<p>Finally, you can run the <code>cmain_*</code> functions used by <code><a href="../reference/pipeline.html">pipeline()</a></code> directly. For example, to run the deconvolution step, you can call <code><a href="../reference/cmain_deconvolve_methylation.html">cmain_deconvolve_methylation()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="my-cna-solution-wasnt-right--how-can-i-refit-with-different-purity-and-ploidy-values">My CNA solution wasn’t right. How can I refit with different purity and ploidy values?<a class="anchor" aria-label="anchor" href="#my-cna-solution-wasnt-right--how-can-i-refit-with-different-purity-and-ploidy-values"></a>
</h3>
<p>If you want to use an external purity and ploidy solution, simply pass a CNA file that has only the purity and ploidy fields. Additionally, set <code>refit==TRUE</code> in the CamConfig and CAMDAC will use this to refit the sample.</p>
</div>
<div class="section level3">
<h3 id="can-i-limit-my-analysis-to-specific-regions-of-interest">Can I limit my analysis to specific regions of interest?<a class="anchor" aria-label="anchor" href="#can-i-limit-my-analysis-to-specific-regions-of-interest"></a>
</h3>
<p>To analyse specific genomic regions, you may pass a BED file to CAMDAC config:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/CamConfig.html">CamConfig</a></span><span class="op">(</span>outdir<span class="op">=</span><span class="st">"."</span>, ref<span class="op">=</span><span class="st">"./pipeline_files"</span>, regions<span class="op">=</span><span class="st">"regions.bed"</span><span class="op">)</span></span></code></pre></div>
<p>CAMDAC will merge any overlapping regions prior to analysis.</p>
</div>
<div class="section level3">
<h3 id="how-can-i-manually-replace-pipeline-outputs">How can I manually replace pipeline outputs?<a class="anchor" aria-label="anchor" href="#how-can-i-manually-replace-pipeline-outputs"></a>
</h3>
<p>If you have outputs from a previous run, you can manually assign them to a CAMDAC object. This overwrites the expected path for that output type, allowing the pipeline to run with this data instead of computing it. Use the <code>attach_output</code> function, passing one of three arguments:</p>
<ul>
<li>
<code>counts</code>: CAMDAC allele counts <code>*.SNP.CpGs.all.sorted.csv.gz</code> file</li>
<li>
<code>snps</code>: CAMDAC sample SNP counts <code>*.SNPs.csv.gz</code> file</li>
<li>
<code>meth</code>: CAMDAC bulk methylation <code>*.m.csv.gz</code> file</li>
<li>
<code>cna</code>: CAMDAC CNA <code>*.cna.txt</code> file</li>
<li>
<code>pure</code>: CAMDAC deconvolved methylation <code>*.m.pure.csv.gz</code> file</li>
</ul>
<p>For example, to attach a previous counts file to a CAMDAC object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CAMDAC</span><span class="op">)</span></span>
<span><span class="va">tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id <span class="op">=</span> <span class="st">"T"</span>, sex <span class="op">=</span> <span class="st">"XY"</span>, bam <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamConfig.html">CamConfig</a></span><span class="op">(</span>outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, build<span class="op">=</span><span class="st">"hg38"</span>, bsseq<span class="op">=</span><span class="st">"wgbs"</span>, lib<span class="op">=</span><span class="st">"pe"</span><span class="op">)</span></span>
<span><span class="va">counts_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"test.SNPs.CpGs.all.sorted.csv.gz"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="va">tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/attach_output.html">attach_output</a></span><span class="op">(</span><span class="va">tumor</span>, <span class="va">config</span>, <span class="st">"counts"</span>, <span class="va">counts_file</span><span class="op">)</span></span></code></pre></div>
<p>The CAMDAC pipeline can now access the file in the expected location at <code>config$outdir</code>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Elizabeth Larose Cadieux, Nana Mensah, Carla Castignani, Jonas Demeulemeester, Peter Van Loo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
